mautic/mautic
# Dynamic web content filters

| Q  | A
| --- | ---
| Bug fix? | N
| New feature? | Y
| Related user documentation PR URL | https://github.com/mautic/documentation/pull/234
| Related developer documentation PR URL | N/A
| Issues addressed (#s or URLs) | N/A
| BC breaks? | N/A
| Deprecations? | N/A

#### Description:
This PR adds a new feature to Dynamic Web Content items: The ability to configure them to be "campaign or filter" based.
Right now, DWC items are only campaign based, but with this new feature the users will be able to configure filters (like in segments) so a Landing Page with a DWC token can show content based on specific conditions.

NOTES:
- This PR fixes an issue with pages increasing in size with each save because the CSS classes to the HTML tag were getting duplicated. 
- This PR also adds a label with the slot type to the slots headers when hovering over them in the builders.

#### Steps to test this PR:
1. Apply this PR
2. Create or edit a Dynamic Web Content item.
3. Notice the "Is campaign based" toggle button.

![image](https://user-images.githubusercontent.com/2924026/31420138-021441d2-adfd-11e7-87c5-6432a0308a69.png)

4. When No is selected, the `Slot Name` field is available and the filters tab is displayed so new filters can be created.

![image](https://user-images.githubusercontent.com/2924026/31420157-22102ed8-adfd-11e7-80ff-6e1eb4696fb9.png)

5. On a landing page, add a token or a slot for the new dynamic web content you just created.

![image](https://user-images.githubusercontent.com/2924026/31420493-1b57f75e-adff-11e7-96c1-e04b6c36a9c8.png)

6. When using DWC slots, you need to provide the requested slot name.

![image](https://user-images.githubusercontent.com/2924026/31421987-113d50bc-ae08-11e7-914a-f526d346f1a1.png)

7. Visit the page and if the conditions on the filters are met, the DWC will be displayed.

![image](https://user-images.githubusercontent.com/2924026/31420536-5e93c994-adff-11e7-9b92-ef63733a4013.png)

8. To test a DWC item with filters in a third party page, you can use the following code:

```
<!DOCTYPE html>
<html>
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
    <title>It works</title>
  </head>
  <body>
    <h1>DWC Test</h1>
<hr/>

<div data-slot="dwc" data-param-slot-name="myslot">
  <h1>Dynamic web content for myslot</h1>
</div>

<script>
    (function(w,d,t,u,n,a,m){w['MauticTrackingObject']=n;
        w[n]=w[n]||function(){(w[n].q=w[n].q||[]).push(arguments)},a=d.createElement(t),
        m=d.getElementsByTagName(t)[0];a.async=1;a.src=u;m.parentNode.insertBefore(a,m)
    })(window,document,'script','//mautic/index_dev.php/mtc.js','mt');

    mt('send', 'pageview');
</script>

  </body>
</html>
```

9. Visit the page and the DWC slot will be replaced if the filters match:

![image](https://user-images.githubusercontent.com/2924026/31420590-acac2392-adff-11e7-8054-8462a82aeb15.png)

10. Make sure the current campaign-based functionality is not affected. See The [Dynamic Web Content documentation](https://mautic.org/docs/en/dwc/index.html) for details on how to test this scenario.
